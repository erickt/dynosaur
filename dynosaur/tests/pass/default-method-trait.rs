use dynosaur::dynosaur;

#[dynosaur(DynMyTrait)]
trait MyTrait {
    fn foo(&mut self) -> impl Send {
        10
    }
}

fn main() {}
