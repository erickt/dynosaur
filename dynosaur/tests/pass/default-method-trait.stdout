#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use dynosaur::dynosaur;

trait MyTrait {
    fn foo(&mut self) -> impl Send { 10 }
}
mod _dynosaur_macro_dynmytrait {
    use super::*;
    trait ErasedMyTrait {
        fn foo<'life0, 'dynosaur>(&'life0 mut self)
        -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>>
        where
        'life0: 'dynosaur,
        Self: 'dynosaur;
    }
    impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
        fn foo<'life0, 'dynosaur>(&'life0 mut self)
            -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>> where
            'life0: 'dynosaur, Self: 'dynosaur {
            Box::pin(<Self as MyTrait>::foo(self))
        }
    }
    #[repr(transparent)]
    pub struct DynMyTrait<'dynosaur_struct> {
        ptr: dyn ErasedMyTrait + 'dynosaur_struct,
    }
    impl<'dynosaur_struct> MyTrait for DynMyTrait<'dynosaur_struct> {
        fn foo(&mut self) -> impl Send {
            let fut: ::core::pin::Pin<Box<dyn Send + '_>> = self.ptr.foo();
            let fut: ::core::pin::Pin<Box<dyn Send + 'static>> =
                unsafe { ::core::mem::transmute(fut) };
            fut
        }
    }
    impl<'dynosaur_struct> DynMyTrait<'dynosaur_struct> {
        pub fn new(value: Box<impl MyTrait + 'dynosaur_struct>)
            -> Box<DynMyTrait<'dynosaur_struct>> {
            let value: Box<dyn ErasedMyTrait + 'dynosaur_struct> = value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn boxed(value: impl MyTrait + 'dynosaur_struct)
            -> Box<DynMyTrait<'dynosaur_struct>> {
            Self::new(Box::new(value))
        }
        pub fn from_ref(value: &(impl MyTrait + 'dynosaur_struct))
            -> &DynMyTrait<'dynosaur_struct> {
            let value: &(dyn ErasedMyTrait + 'dynosaur_struct) = &*value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn from_mut(value: &mut (impl MyTrait + 'dynosaur_struct))
            -> &mut DynMyTrait<'dynosaur_struct> {
            let value: &mut (dyn ErasedMyTrait + 'dynosaur_struct) =
                &mut *value;
            unsafe { ::core::mem::transmute(value) }
        }
    }
}
use _dynosaur_macro_dynmytrait::DynMyTrait;

fn main() {}
